File: dotnet/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example2/OrderCollection.cs
namespace IteratorDesignPattern.Example2;

public class OrderCollection : IAggregate<Order>
{
    private readonly List<Order> _orders = new();

    public void AddOrder(Order order)
    {
        _orders.Add(order);
    }

    public int Count => _orders.Count;

    public Order this[int index] => _orders[index];

    public IIterator<Order> CreateIterator()
    {
        return new OrderIterator(this);
    }

    public IIterator<Order> CreateStatusFilterIterator(OrderStatus status)
    {
        return new StatusFilterIterator(this, status);
    }

    public IIterator<Order> CreateDateRangeIterator(DateTime startDate, DateTime endDate)
    {
        return new DateRangeIterator(this, startDate, endDate);
    }

    public IIterator<Order> CreateHighValueIterator(decimal minAmount)
    {
        return new HighValueOrderIterator(this, minAmount);
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/IteratorDesignPattern/IteratorDesignPattern.Example2/src/OrderCollection.ts
import { IAggregate, IIterator } from "./Interfaces";
import { Order, OrderStatus } from "./Order";
import {
  DateRangeIterator,
  HighValueOrderIterator,
  OrderIterator,
  StatusFilterIterator,
} from "./Iterators";

export class OrderCollection implements IAggregate<Order> {
  private readonly _orders: Order[] = [];

  addOrder(order: Order): void {
    this._orders.push(order);
  }

  get count(): number {
    return this._orders.length;
  }

  getAt(index: number): Order {
    return this._orders[index];
  }

  createIterator(): IIterator<Order> {
    return new OrderIterator(this);
  }

  createStatusFilterIterator(status: OrderStatus): IIterator<Order> {
    return new StatusFilterIterator(this, status);
  }

  createDateRangeIterator(startDate: Date, endDate: Date): IIterator<Order> {
    return new DateRangeIterator(this, startDate, endDate);
  }

  createHighValueIterator(minAmount: number): IIterator<Order> {
    return new HighValueOrderIterator(this, minAmount);
  }
}
